cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -621,11 +621,18 @@ class DatasetWriter::DatasetWriterImpl {
         backpressure = writer_state_.open_files_throttle.Acquire(1);
         if (!backpressure.is_finished()) {
           EVENT_ON_CURRENT_SPAN("DatasetWriter::Backpressure::TooManyOpenFiles");
+          writer_state_.rows_in_flight_throttle.Release(next_chunk->num_rows());
           RETURN_NOT_OK(TryCloseLargestFile());
           break;
         }
       }
-      RETURN_NOT_OK(dir_queue->StartWrite(next_chunk));
+      auto s = dir_queue->StartWrite(next_chunk);
+      if (!s.ok()) {
+        // If `StartWrite` succeeded, it will Release the
+        // `rows_in_flight_throttle` when the write task is finished.
+        writer_state_.rows_in_flight_throttle.Release(next_chunk->num_rows());
```

Comment:
I'm considering should I release `open_files_throttle` here, I gothrough the impl, seems that `OpenFileQueue` would not failed, so I didn't release `open_files_throttle`

Reply:
Could you add the note as a comment?